grpc/src/main/java/com/linecorp/armeria/internal/server/grpc/AbstractServerCall.java

```diff
@@ -246,21 +262,26 @@ private void doClose(Status status, Metadata metadata, @Nullable Throwable excep
                    status, exception);
         closeCalled = true;
 
-        boolean completed = true;
-        if (status.getCode() == Code.CANCELLED && status.getCause() instanceof RequestTimeoutException) {
+        // Override the default behavior of setting "completed == true" if the response
+        // is written successfully by invoking the listener immediately.
+        if (preCloseListener) {
+            // the call is closed due to a reason out of the user's control. (e.g. deframer exception)
+            closeListener(status, metadata, status.isOk(), true);
+        } else if (status.getCode() == Code.CANCELLED && status.getCause() instanceof RequestTimeoutException) {
             // A call was finished by a timeout scheduler, not a user.
-            completed = false;
+            closeListener(status, metadata, false, true);
         } else if (status.isOk() && method.getType().serverSendsOneMessage() && firstResponse() == null) {
             // A call that should send a message incompletely finished.
+            // Preemptively call closeListener and then write the response.
             final String description = "Completed without a response";
             logger.warn("{} {} status: {}, metadata: {}", ctx, description, status, metadata);
             status = Status.CANCELLED.withDescription(description);
-            completed = false;
+            closeListener(status, metadata, false, true);
         }
```

Comment:
The boolean logic and if-else conditions seem complicated. Sorry, I wrote many of them. 😅

We may need to refactor the code before adding more logic. What do you think of extending `StatusAndMetadata` to express more states that would be useful to perform our logic more statically?
`StatusAndMetadata` could be created in advance in the early phase of the closing process. Subsequent actions will override the status of `StatusAndMetadata` if necessary.

For example:
```java
class StatusAndMetadata {
   private final Status status;
   private final Metadata metadata;
   // true if `ServerCall.close()` is triggered by the client
   private final boolean canceled;
   // Set true if response trailers are successfully written
   private boolean completed;
   // Set true if response content is written
   private setResponseContent;
}
// canceled will be will set when is created
StatusAndMetadata statusAndMetadata = ...
if (responseContentIsSet) {
   statusAndMetadata.setResponseContent = true;   
}
if (trailersIsWritten) {
   statusAndMetadata.completed = true;   
}
// `statusMetadata` will be the only argument for closeListener.
closeListener(statusMetadata);
```

If you think that is out of scope of this PR, I'm okay with the current change and let's handle them separately.

Reply:
Reverted optimizations + I just added the extra boolean field like you suggested

Reply:
I wasn't sure my suggestion was more readable but my intention was to unify the management of `canceled` and `completed` in one place to provide better maintainability. However, I didn't prototype my idea so it may be worse.

I'm not strong on my suggestion. If you don't see it is a better approach, feel free to let me know. :-)

Reply:
It seems like `cancelled` and `completed` are only used for deciding to call `invokeOnComplete` or `invokeOnCancel`. How about merging those fields into one (e.g `shouldCancel`)?